### DISCUSSION TITLE
External magnetic field in HybridPIC

### DISCUSSION THREAD

**Discussion by kli-jfp** [BLAST-WarpX/warpx#4899] [Q&A] [answered]
What is the best way to get an external magnetic and/or electric field in the HybridPIC code? If I choose the regular external fields for the particle push then I rightly get an error message. Is this a general problem with HybridPIC codes or is it just not implemented yet? Would this be difficult to implement in the code? 

I'm guessing the easiest way would be to include a static external field B_0 in ohms law so that the total field is B=B_ind+B_0 and curl(B_0)=0.

Thanks!

**Comment by roelof-groenewald** [comment]
Hi @kli-jfp. The equations governing the Ohm's law based hybrid-PIC model are unfortunately not amenable to an external field the way that Maxwell's equations are. However, the model does allow for an external current, which is already included in WarpX. If you want to mimic an external B-field, you can achieve that by specifying an appropriate external current which would result in the desired B-field.
The external current can be specified as a spatial and temporal expression:
```
hybrid_pic_model.J[x/y/z]_external_grid_function(x, y, z, t) = ...
```

**Reply by kli-jfp** [reply]
Ok, I was considering implementing a static external magnetic field in the code, but it would probably be easier to get the external current density from a vector potential. Thanks!

**Comment by roelof-groenewald** [comment]
I also wanted to just make sure you are aware that the initial fields can be set to anything you like via the options described in [this](https://warpx.readthedocs.io/en/latest/usage/parameters.html#applied-to-the-grid) section of the documentation.

**Reply by kli-jfp** [reply]
Great! So if I'm understanding the algorithm correctly:

1. I need an initial magnetic field. For a classical mirror I would need the plasma induced field B_ind and the magnetic field B_0 from the coils. B_tot = B_ind + B_0, I set B_tot as the initial condition via B_ext_grid_init_style .
2. I calculate the current density J_ext from the coils and insert that in the simulation via hybrid_pic_model.J[x/y/z]_external_grid_function(x, y, z, t) = ... (does this require that the coils be in the simulation domain? Might be tricky :/ )

**Reply by roelof-groenewald** [reply, marked as answer]
For a classical mirror I would just set the initial B-field (using the `B_ext_grid_init_style` functionality) to the total B-field (from the coils and any equilibrium plasma currents) and not explicitly specify the coil currents. The coil currents approach is only needed if you want to simulate time varying currents in the coils.

**Reply by kli-jfp** [reply]
Yes, this could work, although the external field would not b static in that case. I'll see if I can implement a static magnetic field in the mean time. The difference might be negligible but could be worth to have it in the code if it works. Thanks for the feedback!
